generator/internal/discovery_resource.cc

```diff
@@ -26,13 +27,117 @@ namespace google {
+  std::string request_resource_field_name;
+  if (verb == "post" || verb == "patch" || verb == "put") {
+    std::string http_body =
+        request_type.json().value("request_resource_field_name", "*");
+    if (http_body != "*") request_resource_field_name = http_body;
+    absl::StrAppend(&http_option,
+                    absl::StrFormat("      body: \"%s\"\n", http_body));
+  }
+  rpc_options.push_back(absl::StrCat(http_option, "    };"));
+
+  std::string operation_scope;
+  std::vector<std::string> params =
+      method_json.value("parameterOrder", std::vector<std::string>{});
+  if (!params.empty()) {
+    if (internal::Contains(params, "zone")) {
+      operation_scope = "ZoneOperations";
+    } else if (internal::Contains(params, "region")) {
+      operation_scope = "RegionOperations";
+    } else if (absl::StrContains(path, "/global/")) {
+      operation_scope = "GlobalOperations";
+    }
+    if (!request_resource_field_name.empty()) {
+      params.push_back(request_resource_field_name);
+    }
+    rpc_options.push_back(
+        absl::StrFormat("    option (google.api.method_signature) = \"%s\";",
+                        absl::StrJoin(params, ",")));
+  }
+
+  if (method_json.contains("response") &&
+      method_json["response"].value("$ref", "") == "Operation") {
+    rpc_options.push_back(
+        absl::StrFormat("    option (google.cloud.operation_service) = \"%s\";",
+                        operation_scope));
+  }
+  return absl::StrJoin(rpc_options, "\n");
+}
+
+StatusOr<std::string> DiscoveryResource::FormatOAuthScopes() const {
+  std::set<std::string> oauth_scopes;
+  auto const& methods = json_.find("methods");
+  for (auto const& iter : *methods) {
+    auto const& scopes = iter.find("scopes");
+    if (scopes == iter.end()) break;
+    for (auto const& s : *scopes) {
+      oauth_scopes.insert(std::string(s));
+    }
+  }
+  if (oauth_scopes.empty()) {
+    return internal::InvalidArgumentError(
+        absl::StrFormat("No OAuth scopes found for service: %s.", name_));
+  }
+  return absl::StrCat("    \"", absl::StrJoin(oauth_scopes, ",\"\n    \""),
+                      "\";\n");
+}
+
+std::string DiscoveryResource::FormatPackageName(
+    std::string const& product_name, std::string const& version) const {
+  return absl::StrJoin(
+      std::make_tuple(std::string("google.cloud.cpp"), product_name,
+                      CamelCaseToSnakeCase(name_), version),
+      ".");
+}
+
+std::string DiscoveryResource::FormatFilePath(
+    std::string const& product_name, std::string const& version,
+    std::string const& output_path) const {
+  return absl::StrJoin({output_path, std::string("google/cloud"), product_name,
+                        CamelCaseToSnakeCase(name_), version,
+                        absl::StrCat(CamelCaseToSnakeCase(name_), ".proto")},
+                       "/");
+}
+
+// TODO(#11138): Add logic to keep resource name plural for aggregatedList and
```

Comment:
I am not even sure this is worth trying. English is too hard. Words lose an `s`, `es`, switch from `a` -> `um`, etc. Yuck.

Here is my laughable attempt: `ListFoos`, `GetOneOfMyFoos` :)